File: bintree_csharp/binarytree/traversal/RecurTraversal.cs
using System;
using System.Collections.Generic;


namespace my.binarytree.traversal
{
    class RecurTraversal<TKey, TNode> : BaseTraversal<TKey, TNode>
        where TKey : IComparable where TNode : BinNode<TKey, TNode>, new()
    {
        protected List<TKey> ResPath = null;



        public override List<TKey> Traverse(TNode root, OrderTraversal order)
        {
            ResPath = new List<TKey>();

            switch (order)
            {
                case OrderTraversal.PRE:
                    TraversePre(root);
                    break;

                case OrderTraversal.IN:
                    TraverseIn(root);
                    break;

                case OrderTraversal.POST:
                    TraversePost(root);
                    break;

                default:
                    break;
            }

            return ResPath;
        }



        protected void TraversePre(TNode node)
        {
            if (null == node)
                return;

            ResPath.Add(node.Key);
            TraversePre(node.Left);
            TraversePre(node.Right);
        }



        protected void TraverseIn(TNode node)
        {
            if (null == node)
                return;

            TraverseIn(node.Left);
            ResPath.Add(node.Key);
            TraverseIn(node.Right);
        }



        protected void TraversePost(TNode node)
        {
            if (null == node)
                return;

            TraversePost(node.Left);
            TraversePost(node.Right);
            ResPath.Add(node.Key);
        }
    }
}

File: bintree_java/src/binarytree/traversal/RecurTraversal.java
package binarytree.traversal;


import binarytree.BinNode;
import java.util.List;
import java.util.ArrayList;



public class RecurTraversal< TKey extends Comparable<? super TKey>,
                             TNode extends BinNode<TKey, TNode>
                           >
             extends BaseTraversal<TKey, TNode>
{


    protected List<TKey> resPath;



    @Override
    public List<TKey> traverse(TNode root, OrderTraversal order) {
        resPath = new ArrayList<>();

        switch (order) {
        case PRE:
            traversePre(root);
            break;

        case IN:
            traverseIn(root);
            break;

        case POST:
            traversePost(root);
            break;

        default:
            break;
        }

        return resPath;
    }



    protected void traversePre(TNode node) {
        if (null == node)
            return;

        resPath.add(node.key);
        traversePre(node.left);
        traversePre(node.right);
    }



    protected void traverseIn(TNode node) {
        if (null == node)
            return;

        traverseIn(node.left);
        resPath.add(node.key);
        traverseIn(node.right);
    }



    protected void traversePost(TNode node) {
        if (null == node)
            return;

        traversePost(node.left);
        traversePost(node.right);
        resPath.add(node.key);
    }
}
